src/player.c

```diff
@@ -2167,11 +2173,89 @@ playback_next_bh(void *arg, int *retval)
   return COMMAND_END;
 }
 
+static int
+seek_calc_position_ms(struct player_seek_param *seek_param, struct db_queue_item **queue_item, int *position_ms)
+{
+  struct db_queue_item *seek_queue_item = NULL;
+  int seek_ms = 0;
+
+  // Initialize out parameters
+  *queue_item = NULL;
+  *position_ms = 0;
+
+  // Calculate seek position
+  if (seek_param->mode == PLAYER_SEEK_POSITION)
+    seek_ms = seek_param->ms;
+  else
+    seek_ms = pb_session.playing_now->pos_ms + seek_param->ms;
+
+  // Check if we need to switch to a previous track, this will be done if we are in the first 3 seconds
+  // of a track and we have a seek request for more than 3 seconds
+  if (seek_ms < 0)
+    {
```

Comment:
The below is pretty clear, but if you want to reduce the if-depth you could, since in the below seek_ms becomes 0 except if pos_ms < 3000 && there is a previous item && the previous item's song_length + seek_ms is positive.